enhancements/microshift/y-minus-2-upgrades.md

```diff
@@ -0,0 +1,304 @@
+  - https://issues.redhat.com/browse/USHIFT-2246
+see-also:
+  - "/enhancements/microshift/microshift-updateability-ostree.md"
+  - "/enhancements/update/eus-upgrades-mvp.md"
+replaces: []
+superseded-by: []
+---
+
+# Upgrading from 4.Y-2 to 4.Y
+
+## Summary
+
+This enhancement describes how MicroShift will support upgrading
+in-place across 2 minor versions at a time.
+
+## Motivation
+
+We are already seeing a tendency for MicroShift users to adopt EUS
+versions and stay on them until they can update to the next EUS
+release. This makes sense given the deployment scenarios for
+MicroShift, which often involve remote locations, limited bandwidth,
+or other reasons that make the appetite for frequent updates as low
+as, or lower than, it is for OpenShift users.
+
+### User Stories
+
+As an edge device administrator, I want to deploy versions of the
+platform software (OS, MicroShift, etc.) with the longest support
+life-cycle so I can focus on my own applications and _using_ the
+device.
+
+As an edge device administrator, I want to upgrade from one
+long-life-cycle version of the platform software directly to another,
+without applying the intermediate version.
+
+### Goals
+
+* Support updating single-node deployments of MicroShift in place on
+  RPM-based and ostree-based systems from version 4.Y-2 to 4.Y.
+
+### Non-Goals
+
+* Multi-node support for MicroShift has been discussed, but is out of
+  scope for this enhancement.
+* Upgrading and skipping versions always requires a full host reboot
+  to ensure all components are restarted and we have no plans to
+  remove that requirement.
+
+## Proposal
+
+Versions 4.12 and 4.13 of MicroShift were preview releases. We did not
+intend to support upgrading to 4.14 from either earlier version at
+all, but did implement upgrade testing as part of preparing 4.14 for
+release. We wanted to limit that testing to 1 version. Therefore, in
+4.14 we introduced an explicit version check to determine if the data
+version (the contents of `/var/lib/microshift` are more than 1 minor
+version older than the software version (the version embedded in the
+new binary). If the skew is too great, MicroShift exits with an error.
+
+To implement this enhancement, we will change the check to support a
+skew of 2 versions.
+
+We expect this to require minimal work in MicroShift because
+
+* The storage migration controller is already running and can be used
+  to update storage versions of any resources.
+* There are not currently any changes to the etcd storage format.
+* The version skew check in MicroShift itself is straightforward to
+  change.
+
+### Workflow Description
+
+1. Edge device administrator deploys a host with MicroShift 4.Y-2
+   installed.
+2. Software runs, time passes.
+3. Edge device administrator updates the host to run MicroShift 4.Y.
+  * For ostree-based systems, the host is automatically rebooted as
```

Comment:
Where would it roll back to if the upgrade fails (for some other reason)? Does ostree have any validation of version skew?

Reply:
The rollback is always to the previous image that was installed on the host, regardless of what the new image contains or what causes the rollback.